nodejs/device/src/index.ts

```diff
@@ -0,0 +1,318 @@
+        res();
+      }).catch(e => rej(e));
+
+      // If the promise takes too long to settle, stop the pairing operation and reject.
+      this.pairingTimer = setTimeout(() => {
+        this.twin.removeAllListeners('properties.desired.thief.pairing');
+        clearTimeout(this.pairingAttemptTimer);
+        rej(new TimeoutError(`No response to pairing requests after trying for ${this.settings.pairingRequestTimeoutIntervalInSeconds} seconds.`));
+      }, 1000 * this.settings.pairingRequestTimeoutIntervalInSeconds);
+    });
+  }
+
+  // Does not deal with out-of-order messages.
+  // For example, if we get index 5 and then 7, we assume 6 is lost.
+  private handleTestC2dMessage(obj, msg) {
+    if (Number.isNaN(obj.thief.testC2dMessageIndex) || typeof obj.thief.testC2dMessageIndex !== 'number') {
+      this.rejectC2d(msg);
+      return console.warn('Issue with testC2dMessageIndex property. Rejecting.');
+    }
+
+    this.completeC2d(msg);
+    if (!this.maxReceivedIndex && this.maxReceivedIndex !== 0) {
+      console.info(`Received initial testC2dMessageIndex: ${this.maxReceivedIndex = obj.thief.testC2dMessageIndex}`);
+    }
+    else if (this.maxReceivedIndex + 1 === obj.thief.testC2dMessageIndex) {
+      console.info(`Received next testC2dMessageIndex: ${++this.maxReceivedIndex}`);
+    }
+    else if (this.maxReceivedIndex < obj.thief.testC2dMessageIndex) {
+      const missing = Array.from({length: obj.thief.testC2dMessageIndex - this.maxReceivedIndex - 1}, (_, i) => 1 + i + this.maxReceivedIndex); 
+      console.warn(`Received testC2dMessageIndex ${obj.thief.testC2dMessageIndex}, but never received indices ${missing}`);
+      this.maxReceivedIndex = obj.thief.testC2dMessageIndex;
+      if ((this.c2dMissingMessageCount += missing.length) > this.settings.receiveC2dAllowedMissingMessageCount) {
+        this.stopDeviceApp(`The number of missing C2D messages exceeds receiveC2dAllowedMissingMessageCount of ${this.settings.receiveC2dAllowedMissingMessageCount}.`);
+      }
+    }
+    else {
+      console.warn(`Received old index: ${obj.thief.testC2dMessageIndex}`);
+    }
+  }
+
+  private handleServiceAckResponseMessage(obj, msg) {
+    if (!Array.isArray(obj.thief.serviceAcks)) {
+      this.rejectC2d(msg);
+      return console.warn('Issue with serviceAcks property. Rejecting.');
+    }
+    console.info(`Received serviceAckIds: ${obj.thief.serviceAcks}`);
+    obj.thief.serviceAcks.forEach(id => {
+      if (this.serviceAckWaitList.has(id)) {
+        this.serviceAckWaitList.delete(id);
+      }
+      else {
+        console.warn(`Received unknown serviceAckId: ${id}`);
+      }
+    });
+    this.completeC2d(msg);
+  }
+
+  private rejectC2d(msg) {
+    this.client.reject(msg).catch(e => {console.warn('Error when rejecting C2D message: ' + e.message)});
+  }
+
+  private completeC2d(msg) {
+    this.client.complete(msg).catch(e => {console.warn('Error when completing C2D message: ' + e.message)});
```

Comment:
>catch(e => {console.warn('Error when completing C2D message: ' + e.message)} [](start = 30, length = 76)

This function has the side-effect of swallowing failures.  If `complete(msg)` throws an exception, it prints something to the log, but nothing else changes.  Maybe that's OK, but I can't say that I'm a big fan of ignoring failures like this.

Reply:
I noticed that too, and I agree. Though we want the device app to continue running when errors like this happen.
What are your thoughts on incrementing a counter and stopping the device app if the counter gets too high? We could add a field to `thiefSettings` where the max value of the counter can be configured.

Reply:
that's exactly what I was thinking

---
In reply to: [577957666](https://github.com/Azure/iot-sdk-longhaul/pull/61#discussion_r577957666) [](ancestors = 577957666)

Reply:
@BertKleewein In the python app we have `THIEF_ALLOWED_CLIENT_LIBRARY_EXCEPTION_COUNT` in the `device_run_config` dict. Should the equivalent configurable value in the Node app be used as a max value here?